src/librustc_mir/interpret/step.rs

```diff
@@ -150,83 +154,89 @@ impl<'a, 'tcx> EvalContext<'a, 'tcx> {
+            // FIXME: check that it's `#[linkage = "extern_weak"]`
             trace!("Initializing an extern global with NULL");
-            self.ecx.globals.insert(cid, Global::initialized(self.ecx.tcx.type_of(def_id), Value::ByVal(PrimVal::Bytes(0)), mutability));
-            return;
+            self.globals.insert(cid, Global::initialized(self.tcx.type_of(def_id), Value::ByVal(PrimVal::Bytes(0)), mutability));
+            return Ok(false);
         }
-        self.try(|this| {
-            let mir = this.ecx.load_mir(instance.def)?;
-            this.ecx.globals.insert(cid, Global::uninitialized(mir.return_ty));
-            let internally_mutable = !mir.return_ty.is_freeze(
-                    this.ecx.tcx,
-                    ty::ParamEnv::empty(Reveal::All),
-                    span);
-            let mutability = if mutability == Mutability::Mutable || internally_mutable {
-                Mutability::Mutable
-            } else {
-                Mutability::Immutable
-            };
-            let cleanup = StackPopCleanup::MarkStatic(mutability);
-            let name = ty::tls::with(|tcx| tcx.item_path_str(def_id));
-            trace!("pushing stack frame for global: {}", name);
-            this.ecx.push_stack_frame(
-                instance,
-                span,
-                mir,
-                Lvalue::Global(cid),
-                cleanup,
-            )
-        });
+        let mir = self.load_mir(instance.def)?;
+        self.globals.insert(cid, Global::uninitialized(mir.return_ty));
+        let internally_mutable = !mir.return_ty.is_freeze(
+                self.tcx,
+                ty::ParamEnv::empty(Reveal::All),
+                span);
+        let mutability = if mutability == Mutability::Mutable || internally_mutable {
+            Mutability::Mutable
+        } else {
+            Mutability::Immutable
+        };
+        let cleanup = StackPopCleanup::MarkStatic(mutability);
+        let name = ty::tls::with(|tcx| tcx.item_path_str(def_id));
+        trace!("pushing stack frame for global: {}", name);
+        self.push_stack_frame(
+            instance,
+            span,
+            mir,
+            Lvalue::Global(cid),
+            cleanup,
+        )?;
+        Ok(true)
     }
+}
 
-    fn try<F: FnOnce(&mut Self) -> EvalResult<'tcx>>(&mut self, f: F) {
-        if let Ok(ref mut n) = *self.new_constants {
-            *n += 1;
-        } else {
-            return;
-        }
-        if let Err(e) = f(self) {
-            *self.new_constants = Err(e);
+// WARNING: make sure that any methods implemented on this type don't ever access ecx.stack
+// this includes any method that might access the stack
+// basically don't call anything other than `load_mir`, `alloc_ptr`, `push_stack_frame`
+// The reason for this is, that `push_stack_frame` modifies the stack out of obvious reasons
+struct ConstantExtractor<'a, 'b: 'a, 'tcx: 'b, M: Machine<'tcx> + 'a> {
```

Comment:
I never really understood this comment. Why does the fact that `push_stack_frame` push a stack frame entail that all methods on this type (?!?) should not access the stack? What exactly goes wrong if they do?

(This doesn't have to be solved by this PR.)